src/sources/kafka.rs

```diff
@@ -370,105 +390,556 @@ impl SourceConfig for KafkaSourceConfig {
 
 async fn kafka_source(
     config: KafkaSourceConfig,
-    consumer: StreamConsumer<CustomContext>,
+    consumer: StreamConsumer<KafkaSourceContext>,
     decoder: Decoder,
-    mut shutdown: ShutdownSignal,
-    mut out: SourceSender,
-    acknowledgements: bool,
+    out: SourceSender,
+    shutdown: ShutdownSignal,
+    eof: bool,
     log_namespace: LogNamespace,
 ) -> Result<(), ()> {
+    let span = info_span!("kafka_source");
     let consumer = Arc::new(consumer);
-    let (finalizer, mut ack_stream) =
-        OrderedFinalizer::<FinalizerEntry>::maybe_new(acknowledgements, Some(shutdown.clone()));
-    let finalizer = finalizer.map(Arc::new);
-    if let Some(finalizer) = &finalizer {
-        consumer
-            .context()
-            .finalizer
-            .set(Arc::clone(finalizer))
-            .expect("Finalizer is only set once");
+
+    consumer
+        .context()
+        .consumer
+        .set(Arc::downgrade(&consumer))
+        .expect("Error setting up consumer context.");
+
+    let (callback_sender, callback_rx) = mpsc::unbounded_channel();
+    consumer
+        .context()
+        .callbacks
+        .set(callback_sender)
+        .expect("Error setting up consumer callback channel.");
+
+    // EOF signal allowing the coordination task to tell the kafka client task when all partitions have reached EOF
+    let (eof_tx, eof_rx) = if eof {
+        let (tx, rx) = oneshot::channel::<()>();
+        (Some(tx), Some(rx))
+    } else {
+        (None, None)
+    };
+
+    let coordination_task = {
+        let span = span.clone();
+        let consumer = Arc::clone(&consumer);
+        let session_timeout_ms = config.session_timeout_ms;
+        let consumer_state =
+            ConsumerStateInner::<Consuming>::new(config, decoder, out, log_namespace);
+        tokio::spawn(async move {
+            let _enter = span.enter();
+            coordinate_kafka_callbacks(
+                consumer,
+                callback_rx,
+                consumer_state,
+                session_timeout_ms,
```

Comment:
I think the drain deadline needs to be less than `session_timeout_ms`. Otherwise, the Kafka broker will presume the client dead when we hit the deadline, right?

Reply:
Yes, that is true; ideally the drain timeout should be a bit shorter to avoid the client getting excluded from the group. 

Without adding any new configuration fields, something like using some percentage of `session_timeout_ms`, e.g. half rounded to the nearest ms, or using `session_timeout_ms - commit_interval_ms`  (assuming the commit interval is less than the session timeout) might be sensible.

I can also see adding a `drain_timeout_ms` configuration option with either a static default or a default based on the session_timeout_ms value. What do you think?

Reply:
I think it would be best to add a `drain_timeout_ms` configuration option with the default being half of `session_timeout_ms` rounded to the nearest ms. This way we can advise uses to tune that parameter if they are seeing undesirable behavior.

Reply:
Sounds good. I didn't see an obvious way to use the serde macros/default functions to initialize this based on a sibling config field, so I made it an `Option<u64>` and calculate the default when setting up this task if `None`. 

Should this emit a warning (or error?) if the drain timeout is set equal or higher than the session timeout?

Reply:
Looks good. I think we should validate the compute drain timeout value and return an error in the build function if it is >= the session timeout, as it is an invariant we should enforce.